### PR TITLE
[ALLUXIO-2829] Add reference counter to the UFS instance in AbstractUfsManager

### PR DESCRIPTION
**Pull request by Reidddddd**
https://alluxio.atlassian.net/browse/ALLUXIO-2829

### CODE REVIEW

**Comment by Reidddddd** [comment]
alluxio-bot, check this please

File: core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java [hunk 1]
```diff
@@ -76,8 +79,12 @@ public String toString() {
     }
   }
 
-  // TODO(binfan): Add refcount to the UFS instance. Once the refcount goes to zero,
-  // we could close this UFS instance.
+  /**
+   * This map keeps track of the reference counter for an {@link UnderFileSystem}.
+   */
+  private final ConcurrentHashMap<Class<? extends UnderFileSystem>, AtomicLong> mReferenceCounter =
```
Comment (apc999) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: the map should be keyed by `Key`, similar to `mUnderFileSystemMap` rather than `Class<? extends UnderFileSystem>`. The reason is, each `UnderFileSystem` is initialized with possibly different configuration after "ALLUXIO-2743". So even two under file systems share the same class, e.g., both are S3UnderFileSystem, they could be totally different as they use different credentials to access S3 buckets. Thus we only want to reuse the UFS when they are exactly the same scheme/authority and also using the same configuration to create, which is represented by class `Key`

Comment (apc999) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: One possible easy way to do so is to have `mUnderFileSystemMap` mapping from `Key` to a tupe of `<UnderFileSystem, AtomicLong>`

**Comment by Reidddddd** [comment]
I'm not sure where cause "Alluxio Tests" fails? It's kind if you can point out, i will fix it, thanks in advance. @apc999

**Comment by apc999** [comment]
Jenkins, test this please

File: core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java [hunk 2]
```diff
@@ -149,22 +159,24 @@ public UnderFileSystem addMount(long mountId, String ufsUri, Map<String, String>
...
   @Override
   public void removeMount(long mountId) {
     Preconditions.checkArgument(mountId != IdUtils.INVALID_MOUNT_ID, "mountId");
-    // TODO(binfan): check the refcount of this ufs in mUnderFileSystemMap and remove it if this is
-    // no more used. Currently, it is possibly used by out mount too.
-    mMountIdToUnderFileSystemMap.remove(mountId);
+    Pair<UnderFileSystem, AtomicLong> ufsCounter = mMountIdToUnderFileSystemMap.get(mountId);
+    // Remove ufs from mMountIdToUnderFileSystemMap if its counter comes to zero.
+    if (ufsCounter.getSecond().decrementAndGet() == 0L) {
+      mMountIdToUnderFileSystemMap.remove(mountId);
```
Comment (apc999) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: more importantly, close the ufs:
`ufsCounter.getFirst().close`

File: core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java [hunk 3]
```diff
@@ -149,22 +159,24 @@ public UnderFileSystem addMount(long mountId, String ufsUri, Map<String, String>
...
   }
 
   @Override
   public void removeMount(long mountId) {
     Preconditions.checkArgument(mountId != IdUtils.INVALID_MOUNT_ID, "mountId");
-    // TODO(binfan): check the refcount of this ufs in mUnderFileSystemMap and remove it if this is
-    // no more used. Currently, it is possibly used by out mount too.
-    mMountIdToUnderFileSystemMap.remove(mountId);
+    Pair<UnderFileSystem, AtomicLong> ufsCounter = mMountIdToUnderFileSystemMap.get(mountId);
+    // Remove ufs from mMountIdToUnderFileSystemMap if its counter comes to zero.
```
Comment (apc999) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: we should always have `mMountIdToUnderFileSystemMap.remove(mountId);` regardless of counter.
but when counter goes to zero, remove it from `mUnderFileSystemMap`

File: core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java [hunk 4]
```diff
@@ -149,22 +159,24 @@ public UnderFileSystem addMount(long mountId, String ufsUri, Map<String, String>
...
+    if (ufsCounter.getSecond().decrementAndGet() == 0L) {
+      mMountIdToUnderFileSystemMap.remove(mountId);
+    }
   }
 
   @Override
   public UnderFileSystem get(long mountId) {
-    return mMountIdToUnderFileSystemMap.get(mountId);
+    return mMountIdToUnderFileSystemMap.get(mountId).getFirst();
```
Comment (apc999) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: `mUnderFileSystemMap.get(mountId)` may return `null`? in that case NPE is thrown

File: core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java (hunk 2 above)
Comment (Reidddddd) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: ufs is already registered in closer `mCloser.register(ufsCounter.getFirst());` just want to make sure if we should explicitly call close().

File: core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java (hunk 3 above)
Comment (Reidddddd) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: To find the ufs in `mUnderFileSystemMap`, do we need to add a map to keep track of mountId -> key? otherwise, iterate whole `mUnderFileSystemMap` to find the ufs? This doubt stuck me long.

File: core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java (hunk 4 above)
Comment (Reidddddd) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: There's a doubt here, `UnderFileSystem get(long mountId)` should count as a one more reference?

File: core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java (hunk 3 above)
Comment (apc999) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: I am fine to iterate the entire `mUnderFileSystemMap` to find the ufs. in reality the number of different ufses stored in the map shouldn't be large. And `removeMount` shouldn't be heavily used too as it is invoked by `unmount`.

File: core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java (hunk 2 above)
Comment (apc999) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: yeah, that's the main purpose to keep refcount --- avoid keeping long-living UFS client. Once a UFS client is not used, close it.

File: core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java (hunk 4 above)
Comment (apc999) [inline core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java]: maybe I am not clear here. Let me rephrase:
in case `mountId` doesn't exist at all in `mMountIdToUnderFileSystemMap`, original implementation will return null as this map contains no mapping for the key, and this method will return null. While the current way changes this behavior: if `mountId` doesn't exist, `mMountIdToUnderFileSystemMap.get(mountId)` will return null still, but `null.getFirst()` will get a `NullPointerException`. So the way to keep the same behavior can be something like this:

```java
Pair<UnderFileSystem, AtomicLong> counter = mMountIdToUnderFileSystemMap.get(mountId);
if (counter == null) {
  return null;
}
return counter.getFirst();
```

**Comment by apc999** [comment]
Explained my comments. Please take a look if they make sense. @Reidddddd

**Comment by Reidddddd** [comment]
Jenkins, test this please

**Comment by yupeng9** [comment]
@apc999 does the latest PR address your comments?
@Reidddddd can you resolve the conflicts?

**Comment by yupeng9** [comment]
@Reidddddd Feel free to reopen this PR if you have updates
